banyand/liaison/grpc/discovery.go

```diff
@@ -234,7 +234,9 @@ func (e *entityRepo) OnAddOrUpdate(schemaMetadata schema.Metadata) {
 	case schema.KindTrace:
 		trace := schemaMetadata.Spec.(*databasev1.Trace)
 		id = getID(trace.GetMetadata())
+		e.RWMutex.Lock()
```

Comment:
It is good but does not cover all scenarios: e.traceIDIndexMap should be protected by a mutex.